### DISCUSSION TITLE
map<string, string> fails to parse

### DISCUSSION THREAD

**Discussion by RReverser** [protobuf-c/protobuf-c#652] [Q&A] [answered]
Having the following message:

``` proto
message StartTag {
  required string name = 1;
  required map<string, string> attributes = 2;
  required bool self_closing = 3;
}
```

I'm getting a compilation error when running `protoc-c` against it: `Expected field name.`.

Version of the tool:

```
protobuf-c 1.2.1
libprotoc 2.6.1
```

**Comment by zapstar** [comment]
Looks like the error message is getting propagated from the underlying protobuf parser and not protobuf-c?

I think this feature is available only with v3? See this [link](http://stackoverflow.com/a/29646245/1084879).

One of the project members should be able to answer whether the support is added for this in protobuf-c compiler as well..

```
# cat hello.proto
message StartTag {
  required string name = 1;
  required map<string, string> attributes = 2;
  required bool self_closing = 3;
}
# protoc --version
libprotoc 2.6.1
# protoc --cpp_out=. hello.proto
hello.proto:3:15: Expected field name.
# protoc --python_out=. hello.proto
hello.proto:3:15: Expected field name.
```

**Comment by RReverser** [comment]
> I think this feature is available only with v3? See this link.

No, it's part of protobuf v2.

https://developers.google.com/protocol-buffers/docs/reference/proto2-spec#map_field

**Comment by RReverser** [comment]
As for the details - I'm not familiar with the structure of this project and whether it's a separate parser or not, just reported an issue here :)

**Comment by randylayman** [comment]
@RReverser did you ever find a resolution to this? I'm running into a similar issue.

**Comment by RReverser** [comment]
@randylayman Nope, had to use separate `.proto` specifically for `protobuf-c`.

As per https://developers.google.com/protocol-buffers/docs/proto#backwards-compatibility `map<A, B>` is basically `repeated Item item` where `Item` is `message Item { A key = 1; B value = 2; }`, so those are binary compatible types.

I would really love to avoid that workaround though and have this type recognized by `protobuf-c` on its own.

**Comment by lipnitsk** [comment, marked as answer]
This seems to work just fine in latest [code](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L504), with no extra work needed in protobuf-c.
```c
  // For maps fields:
  //     map<KeyType, ValueType> map_field = 1;
  // The parsed descriptor looks like:
  //     message MapFieldEntry {
  //         option map_entry = true;
  //         optional KeyType key = 1;
  //         optional ValueType value = 2;
  //     }
  //     repeated MapFieldEntry map_field = 1;
```
